src/class_decorator_downlevel_transformer.ts

```diff
@@ -0,0 +1,119 @@
+import * as ts from 'typescript';
+
+import {DecoratorClassVisitor} from './decorator-annotator';
+
+function createClassDecoratorType() {
+  const typeElements: ts.TypeElement[] = [];
+  typeElements.push(ts.createPropertySignature(
+      undefined, 'type', undefined,
+      ts.createTypeReferenceNode(ts.createIdentifier('Function'), undefined), undefined));
+  typeElements.push(ts.createPropertySignature(
+      undefined, 'args', ts.createToken(ts.SyntaxKind.QuestionToken),
+      ts.createArrayTypeNode(ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)), undefined));
+  return ts.createArrayTypeNode(ts.createTypeLiteralNode(typeElements));
+}
+
+function extractMetadataFromSingleDecorator(decorator: ts.Decorator) {
+  const metadataProperties: ts.ObjectLiteralElementLike[] = [];
+  const expr = decorator.expression;
+  switch (expr.kind) {
+    case ts.SyntaxKind.Identifier:
+      // The decorator was a plain @Foo.
+      metadataProperties.push(ts.createPropertyAssignment('type', expr));
+      break;
+    case ts.SyntaxKind.CallExpression:
+      // The decorator was a call, like @Foo(bar).
+      const call = expr as ts.CallExpression;
+      metadataProperties.push(ts.createPropertyAssignment('type', call.expression));
+      if (call.arguments.length) {
+        const args: ts.Expression[] = [];
+        for (const arg of call.arguments) {
+          args.push(arg);
+        }
+        const argsArrayLiteral = ts.createArrayLiteral(args);
+        argsArrayLiteral.elements.hasTrailingComma = true;
+        metadataProperties.push(ts.createPropertyAssignment('args', argsArrayLiteral));
+      }
+      break;
+    default:
+      throw new Error(`Unimplemented kind: ${ts.SyntaxKind[expr.kind]}`);
+  }
+  return ts.createObjectLiteral(metadataProperties);
+}
+
+function createDecoratorClassProperty(decoratorList: ts.ObjectLiteralExpression[]) {
+  const modifier = ts.createToken(ts.SyntaxKind.StaticKeyword);
+  const type = createClassDecoratorType();
+  const initializer = ts.createArrayLiteral(decoratorList, true);
+  initializer.elements.hasTrailingComma = true;
+  return ts.createProperty(undefined, [modifier], 'decorators', undefined, type, initializer);
+}
+
+function isNameEqual(classMember: ts.ClassElement, name: string): boolean {
+  if (classMember.name === undefined) {
+    return false;
+  }
+  const id = classMember.name as ts.Identifier;
+  return id.text === name;
+}
+
+function insertBeforeDecoratorProperties(
+    classMembers: ts.NodeArray<ts.ClassElement>, decoratorMetadata: ts.PropertyDeclaration) {
+  const ctorParameterIndex = classMembers.findIndex(m => isNameEqual(m, 'ctorParameters'));
```

Comment:
why not just search for isNameEqual('ctorParams') || isNameEqual('propDecorators'), then use that index or 0?